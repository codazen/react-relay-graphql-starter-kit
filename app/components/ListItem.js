/* @flow */

import React from 'react';

type Props = {
  content: string, // initial item content
  id: any, // id of item
  updateItem: ?Function, // callback for updating listItem
  removeItem: ?Function, // callback for removing listItem
};

type State = {
  editStatus: boolean, // keeps track of what view to display
  content: string, // display content of item
  id: any, // id reference for item
};

/**
 * Generic list item component
 * Parent component: List
 * Child component: N/A
 */
export default class ListItem extends React.Component {

  constructor(props: Props) {
    super(props);
    this.state = {
      editStatus: false,
      content: props.content,
      id: props.id,
    };
  }

  state: State;

  props: Props;

  handleChange = (event: any) => {
    const newState = {};
    newState[event.target.name] = event.target.value;
    this.setState(newState);
  }

  /**
   * Makes removeItem call
   */
  handleRemove = () => {
    const {
      removeItem,
    } = this.props;
    if (removeItem) {
      removeItem(this.state.id);
    }
  }

  /**
   * Changes span to input for inline editing an item's content
   */
  openEditMode = () => {
    this.setState({
      editStatus: true,
    });
  }

  /**
   * Makes updateItem call on click out of ListItem through onBlur event listener
   */
  closeEditMode = () => {
    const {
      updateItem,
    } = this.props;
    if (updateItem) {
      updateItem({
        id: this.state.id,
        content: this.state.content,
      });
    }
    this.setState({
      editStatus: false,
    });
  }

  render() {
    return (
      <div className="list-item-component">
        {this.state.editStatus ?
          <input
            autoFocus
            type="text"
            value={this.state.content}
            name="content"
            onChange={this.handleChange}
            onBlur={this.closeEditMode}
          />
          :
          <div>
            <span onClick={this.props.updateItem ? this.openEditMode : null}>
              {this.state.content}
            </span>
            <button onClick={this.handleRemove}>
              x
            </button>
          </div>
        }
      </div>
    );
  }
}
