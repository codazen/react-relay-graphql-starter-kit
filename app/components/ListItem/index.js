/* @flow */

import React from 'react';

type Props = {
  content: string, // initial item content
  id: any, // id of item
  updateItem: ?Function, // eslint-disable-line
  removeItem: ?Function, // eslint-disable-line
};

type State = {
  editStatus: boolean, // keeps track of what view to display
  content: string, // display content of item
  id: any, // id reference for item
};

/**
 * Generic list item component
 * Parent component: List
 * Child component: N/A
 */
export default class ListItem extends React.Component {

  static defaultProps: Props;

  constructor(props: Props) {
    super(props);
    this.state = {
      editStatus: false,
      content: props.content,
      id: props.id,
    };

    (this:any).openEditMode = this.openEditMode.bind(this);
    (this:any).closeEditMode = this.closeEditMode.bind(this);
    (this:any).handleChange = this.handleChange.bind(this);
    (this:any).handleRemove = this.handleRemove.bind(this);
  }

  state: State;

  props: Props;

  handleChange(event: any) {
    const newState = {};
    newState[event.target.name] = event.target.value;
    this.setState(newState);
  }

  handleRemove() {
    const {
      removeItem,
    } = this.props;
    if (removeItem) {
      removeItem(this.state.id);
    }
  }

  openEditMode() {
    this.setState({
      editStatus: true,
    });
  }

  closeEditMode() {
    const {
      updateItem,
    } = this.props;
    if (updateItem) {
      updateItem({
        id: this.state.id,
        content: this.state.content,
      });
    }
    this.setState({
      editStatus: false,
    });
  }

  render() {
    return (
      <div>
        {this.state.editStatus ?
          <div>
            <input
              autoFocus
              type='text'
              value={this.state.content}
              name='content'
              onChange={this.handleChange}
              onBlur={this.closeEditMode}
            />
            <button onClick={this.closeEditMode}>
              Save
            </button>
          </div>
          :
          <div>
            <span onClick={this.props.updateItem ? this.openEditMode : null}>
              {this.state.content}
            </span>
            <button onClick={this.handleRemove}>
              Delete
            </button>
          </div>
        }
      </div>
    );
  }
}
