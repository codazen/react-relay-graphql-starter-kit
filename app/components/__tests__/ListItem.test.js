import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import ListItem from '../ListItem';

// setup React as global variable
global.React = React;

describe('ListItem component', () => {
  let listItem;
  let event;

  // mock props
  const removeItem = jest.fn();
  const updateItem = jest.fn();
  const id = 'testId';
  const content = 'testContent';

  describe('Tests with all defined props', () => {
    beforeEach(() => {
      listItem = shallow(
        <ListItem
          id={id}
          content={content}
          removeItem={removeItem}
          updateItem={updateItem}
        />);
      removeItem.mockClear();
      updateItem.mockClear();
    });

    it('Renders correctly', () => {
      expect(shallowToJson(listItem)).toMatchSnapshot();
    });

    describe('Unit test: handleRemove', () => {
      it('Calls removeItem in handleRemove', () => {
        listItem.instance().handleRemove();
        expect(removeItem).toHaveBeenCalled();
      });
    });

    describe('Unit test: handleChange', () => {
      beforeEach(() => {
        event = {
          target: {
            name: 'content',
            value: 'testValue',
          },
        };
      });

      it('handles input change', () => {
        expect(listItem.instance().state.content).toBe(content);
        listItem.instance().handleChange(event);
        expect(listItem.instance().state.content).toBe('testValue');
      });
    });

    describe('Unit test: openEditMode', () => {
      it('correctly sets state', () => {
        expect(listItem.instance().state.editStatus).toBe(false);
        listItem.instance().openEditMode();
        expect(listItem.instance().state.editStatus).toBe(true);
      });
    });

    describe('Unit test: closeEditMode', () => {
      it('calls updateItem', () => {
        listItem.instance().closeEditMode();
        expect(updateItem).toHaveBeenCalled();
      });

      it('correctly sets state', () => {
        listItem.instance().closeEditMode();
        expect(listItem.instance().state.editStatus).toBe(false);
      });
    });

    describe('removeItem functionality', () => {
      it('Calls removeItem on button click', () => {
        listItem.find('button').simulate('click');
        expect(removeItem).toHaveBeenCalled();
      });
    });

    describe('updateItem functionality', () => {
      it('updateItem available for call', () => {
        expect(listItem.find('span').prop('onClick')).not.toBe(null);
      });

      it('Correctly handles span click, input blur', () => {
        expect(listItem.instance().state.editStatus).toBe(false);
        listItem.find('span').simulate('click');
        expect(listItem.instance().state.editStatus).toBe(true);
        listItem.find('input').simulate('blur');
        expect(updateItem).toHaveBeenCalled();
        expect(listItem.instance().state.editStatus).toBe(false);
      });
    });
  });

  describe('Tests with undefined updateItem prop', () => {
    beforeEach(() => {
      listItem = shallow(
        <ListItem
          id={id}
          content={content}
          removeItem={removeItem}
        />);
      removeItem.mockClear();
      updateItem.mockClear();
    });

    it('Renders correctly', () => {
      expect(shallowToJson(listItem)).toMatchSnapshot();
    });

    describe('Unit test: handleRemove', () => {
      it('Calls removeItem in handleRemove', () => {
        listItem.instance().handleRemove();
        expect(removeItem).toHaveBeenCalled();
      });
    });

    describe('Unit test: handleChange', () => {
      beforeEach(() => {
        event = {
          target: {
            name: 'content',
            value: 'testValue',
          },
        };
      });

      it('handles input change', () => {
        expect(listItem.instance().state.content).toBe(content);
        listItem.instance().handleChange(event);
        expect(listItem.instance().state.content).toBe('testValue');
      });
    });

    describe('Unit test: openEditMode', () => {
      it('correctly sets state', () => {
        expect(listItem.instance().state.editStatus).toBe(false);
        listItem.instance().openEditMode();
        expect(listItem.instance().state.editStatus).toBe(true);
      });
    });

    describe('Unit test: closeEditMode', () => {
      it('doesn\'t call updateItem', () => {
        listItem.instance().closeEditMode();
        expect(updateItem).not.toHaveBeenCalled();
      });

      it('correctly sets state', () => {
        listItem.instance().closeEditMode();
        expect(listItem.instance().state.editStatus).toBe(false);
      });
    });

    describe('removeItem functionality', () => {
      it('Calls removeItem on button click', () => {
        listItem.find('button').simulate('click');
        expect(removeItem).toHaveBeenCalled();
      });
    });

    describe('updateItem functionality', () => {
      it('updateItem is not available for call', () => {
        expect(listItem.find('span').prop('onClick')).toBe(null);
      });

      it('Correctly handles span click', () => {
        expect(listItem.instance().state.editStatus).toBe(false);
        listItem.find('span').simulate('click');
        expect(listItem.instance().state.editStatus).toBe(false);
      });
    });
  });

  describe('Tests with undefined removeItem prop', () => {
    beforeEach(() => {
      listItem = shallow(
        <ListItem
          id={id}
          content={content}
          updateItem={updateItem}
        />);
      removeItem.mockClear();
      updateItem.mockClear();
    });

    it('Renders correctly', () => {
      expect(shallowToJson(listItem)).toMatchSnapshot();
    });

    describe('Unit test: handleRemove', () => {
      it('Doesn\'t call removeItem in handleRemove', () => {
        listItem.instance().handleRemove();
        expect(removeItem).not.toHaveBeenCalled();
      });
    });

    describe('Unit test: handleChange', () => {
      beforeEach(() => {
        event = {
          target: {
            name: 'content',
            value: 'testValue',
          },
        };
      });

      it('handles input change', () => {
        expect(listItem.instance().state.content).toBe(content);
        listItem.instance().handleChange(event);
        expect(listItem.instance().state.content).toBe('testValue');
      });
    });

    describe('Unit test: openEditMode', () => {
      it('correctly sets state', () => {
        expect(listItem.instance().state.editStatus).toBe(false);
        listItem.instance().openEditMode();
        expect(listItem.instance().state.editStatus).toBe(true);
      });
    });

    describe('Unit test: closeEditMode', () => {
      it('calls updateItem', () => {
        listItem.instance().closeEditMode();
        expect(updateItem).toHaveBeenCalled();
      });

      it('correctly sets state', () => {
        listItem.instance().closeEditMode();
        expect(listItem.instance().state.editStatus).toBe(false);
      });
    });

    describe('removeItem functionality', () => {
      it('Doesn\'t call removeItem on button click', () => {
        listItem.find('button').simulate('click');
        expect(removeItem).not.toHaveBeenCalled();
      });
    });

    describe('updateItem functionality', () => {
      it('updateItem is available for call', () => {
        expect(listItem.find('span').prop('onClick')).not.toBe(null);
      });

      it('Correctly handles span click, input blur', () => {
        expect(listItem.instance().state.editStatus).toBe(false);
        listItem.find('span').simulate('click');
        expect(listItem.instance().state.editStatus).toBe(true);
        listItem.find('input').simulate('blur');
        expect(updateItem).toHaveBeenCalled();
        expect(listItem.instance().state.editStatus).toBe(false);
      });
    });
  });
});
