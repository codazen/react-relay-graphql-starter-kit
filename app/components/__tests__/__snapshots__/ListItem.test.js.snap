exports[`ListItem component Tests with all defined props Renders correctly 1`] = `
<div
  className="list-item-component">
  <div>
    <span
      onClick={[Function]}>
      testContent
    </span>
    <button
      onClick={[Function]}>
      x
    </button>
  </div>
</div>
`;

exports[`ListItem component Tests with undefined removeItem prop Renders correctly 1`] = `
<div
  className="list-item-component">
  <div>
    <span
      onClick={[Function]}>
      testContent
    </span>
    <button
      onClick={[Function]}>
      x
    </button>
  </div>
</div>
`;

exports[`ListItem component Tests with undefined updateItem prop Renders correctly 1`] = `
<div
  className="list-item-component">
  <div>
    <span
      onClick={null}>
      testContent
    </span>
    <button
      onClick={[Function]}>
      x
    </button>
  </div>
</div>
`;
